canonical-web-and-design/maas-ui
# Implement machine list grouping and sorting with double row headers

## Done
- Added a TableHeader component which allows table headers to be clickable to run their own sort functions. Used in MachineList and UserList in settings. This required a bit of custom work to override the default MainTable sort behaviour, so I'm not sure this work should be upstreamed to the component library (if ever). I think once the UX of a generic table between products is finalised we can settle on an implementation, but for now this implements what's in the current angular version.
- Added a GroupSelect component that implements grouping as in the current angular version, plus added options for pool, power state and zone.
- Added placeholder filter dropdown and searchbox.
- Refactored MachineList to better separate grouping, sorting and row generation functions.
- Clicking the FQDN | MAC headers now change the first cell content.
- Drive by styling fixes:
  - Dense table headers (should be released in Vanilla soon)
  - Table cell/header spacing
  - Ensure table widths add up to 100% (oops)

## QA
- Check that you can change the leftmost column to MAC or FQDN.
- Check that clicking the table headers changes the sort parameter, and can be cycled between ascending, descending and none. Check that sorting works for all column headers, including FQDN and MAC.
- Check that you can change the grouping of machines, and that changing the sort changes the order of machines within each group (not the groups themselves)
- Check that the Settings > Users table still sorts as expected.

## Fixes
Fixes canonical-web-and-design/MAAS-squad#1744
Fixes canonical-web-and-design/MAAS-squad#1745
Fixes canonical-web-and-design/MAAS-squad#1746

## Screenshot
![0 0 0 0_8400_MAAS_r_machines (1)](https://user-images.githubusercontent.com/25733845/73351878-392a0380-4290-11ea-8501-e13cac32dd60.png)
